crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use approx::abs_diff_eq;
+use hashbrown::hash_map::Iter;
+use hashbrown::{HashMap, HashSet};
+use ndarray::linalg::kron;
+use ndarray::Array2;
+use num_bigint::BigInt;
+use num_complex::Complex64;
+use numpy::PyReadonlyArray2;
+use pyo3::intern;
+use smallvec::SmallVec;
+
+use crate::unitary_compose::compose;
+use pyo3::prelude::*;
+use pyo3::types::{PyBool, PyDict, PySequence, PyString, PyTuple};
+use qiskit_circuit::bit_data::BitData;
+use qiskit_circuit::circuit_instruction::CircuitInstruction;
+use qiskit_circuit::dag_node::DAGOpNode;
+use qiskit_circuit::imports::QI_OPERATOR;
+use qiskit_circuit::operations::OperationRef::{Gate as PyGateType, Operation as PyOperationType};
+use qiskit_circuit::operations::{Operation, OperationRef, Param};
+use qiskit_circuit::{Clbit, Qubit};
+use rustworkx_core::distancemap::DistanceMap;
+
+const SKIPPED_NAMES: [&str; 4] = ["measure", "reset", "delay", "initialize"];
+const NO_CACHE_NAMES: [&str; 2] = ["annotated", "linear_function"];
+const SUPPORTED_OP: [&str; 18] = [
+    "h", "x", "y", "z", "sx", "sxdg", "t", "tdg", "s", "sdg", "cx", "cy", "cz", "swap", "iswap",
+    "ecr", "ccx", "cswap",
+];
+
+#[pyclass(module = "qiskit._accelerate.commutation_checker")]
+struct CommutationChecker {
+    library: CommutationLibrary,
+    cache_max_entries: usize,
+    cache: HashMap<(String, String), CommutationCacheEntry>,
+    #[pyo3(get)]
+    current_cache_entries: usize,
+    #[pyo3(get)]
+    _cache_miss: usize,
+    #[pyo3(get)]
+    _cache_hit: usize,
+    #[pyo3(get)]
+    gates: Option<HashSet<String>>,
+}
+
+#[pymethods]
+impl CommutationChecker {
+    #[pyo3(signature = (standard_gate_commutations=None, cache_max_entries=1_000_000, gates=None))]
+    #[new]
+    fn py_new(
+        py: Python,
+        standard_gate_commutations: Option<Py<PyAny>>,
+        cache_max_entries: usize,
+        gates: Option<HashSet<String>>,
+    ) -> Self {
```

Comment:
We could change this function signature to accept an `Option<Bound<PyAny>>` and skip the `py` here.

```suggestion
    fn py_new(
        standard_gate_commutations: Option<Bound<PyAny>>, // Send a bound here
        cache_max_entries: usize,
        gates: Option<HashSet<String>>,
    ) -> Self {
```